### DISCUSSION TITLE
custom sub has multiple problems in multiplayer (inconsistent signal component behavior)

### DISCUSSION THREAD

**Discussion by Venyll** [FakeFishGames/Barotrauma#13621] [Bug reports]
### Disclaimers

- [X] I have searched the issue tracker to check if the issue has already been reported.
- [X] My issue happened while using mods.

### What happened?

I created a custom submarine to play with friends, it works as intended in the editor and in singleplayer campaign. In multiplayer mode however, when I host a private game, the submarine won't move up, down, left or right, despite the motors and pumps apparently working (if I interact with any of the motors or pumps, they show the thrust settings and pump speeds according to the intended movement and you can hear the sound).
The docking hatches seem to open and close randomly while docked when they would stay open in singleplayer and editor and the navterminal buttons are also not working in the multiplayer session only.
![20240228115821_1](https://github.com/FakeFishGames/Barotrauma/assets/161589767/e4a8f823-2a4c-41b0-a97e-57899d6414e2)

### Reproduction steps

[Squidshark.zip](https://github.com/FakeFishGames/Barotrauma/files/14433938/Squidshark.zip)
For steering use the statusmonitor in the commandroom (monitor is invisible and located on the sonarmonitor above the chair, near the battery charge display). Reactor status is "off" at the start of the game. To activate the motors, use the button on the motor symbol, which is located on the right side of the battery charge display. The motor symbol will glow light blue if active.

### Bug prevalence

everytime in multiplayer only

### Single player or multiplayer?

multiplayer

### Version

v1.2.8.0 (Winter Update hotfix 2)

### Which operating system did you encounter this bug on?

Windows

### Relevant error messages and crash reports

none

**Comment by itchyOwl** [comment]
QA: Check if it's easy to repro the issue and write down the steps. The focus here should be on whether there's some bug in the game that's causing the sub to work incorrectly. If it's not easy to quickly repro, write that down here.

**Comment by Venyll** [comment]
The problems seem to occur everytime in a multiplayer game, no matter if I play alone, with friends, on a public or private server or what missions I take. We have played before with vanilla and modded subs, they all worked fine. I was not able to reproduce these issues when I removed all circuitry except for the circuitry for pumps, motors and terminals or with other subs.
To reproduce the issues, just start a multiplayer game, undock from the station if the sub is currently docked, activate the motors and try to move the sub or try using the navterminals buttons.

**Comment by Regalis11** [comment]
I took a quick look at this in an attempt to diagnose the issue, but I'm afraid I could not find the problem. :/ It seems to have something to do with the network of relays that pass power to the pumps - in multiplayer, the engines don't seem to be receiving any power, despite the button working fine and enabling the relay it's connected to. I would suspect there's some kind of time-sensitivity issue in how the various logic components control the states of the relays (order in which the signals propagate affects the behavior of the circuit?), but this circuit is so enormously complex I don't think there's any hope for me to even understand what's happening here (and even if there was, diagnosing it would take so much time it's probably not justified for an issue that seems like an extremely rare edge case).
![image](https://github.com/FakeFishGames/Barotrauma/assets/14351070/a74a2833-2f66-4854-a53d-157e7601b502)

I'm sorry I'm not able to offer you any more help with this at the moment. It's a really impressive sub, with some really impressive work on the electrical system, so it's a shame the circuit doesn't work properly in multiplayer.

But we can keep the ticket open and hope we can get some more clues on this, or perhaps someone finds a simpler circuit in which the error occurs and is easier to diagnose.
